lib/mix/lib/mix/tasks/do.ex

```diff
@@ -22,29 +24,21 @@ defmodule Mix.Tasks.Do do
     end
   end
 
-  @doc false
-  def gather_commands(args) do
-    gather_commands(args, [], [])
+  defp gather_commands(args) do
+    gather_commands args, [], []
   end
 
-  defp gather_commands([], current, commands) do
-    [current | commands]
-    |> Enum.reject(&(&1 == []))
-    |> Enum.map(&Enum.reverse(&1))
-    |> Enum.reverse
+  defp gather_commands([h|t], current, acc) when binary_part(h, byte_size(h), -1) == "," do
+    part    = binary_part(h, 0, byte_size(h) - 1)
+    current = Enum.reverse([part | current])
+    gather_commands t, [], [current | acc]
   end
 
-  defp gather_commands([arg | rest], current, commands) do
-    case String.split(arg, ",", parts: 2) do
-      [arg] ->
-        gather_commands(rest, [arg | current], commands)
-      [left, right] ->
-        rest    = append_unless_empty(right, rest)
-        current = append_unless_empty(left, current)
-        gather_commands(rest, [], [current | commands])
-    end
+  defp gather_commands([h | t], current, acc) do
```

Comment:
Since we're changing this, I think better to replace one-letter variable with name from previous implementation.